fix(PaginationTableDemo): add Spinner and empty state to demo

nicolethoen opened the pull request:
Fixes #3184

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-3294.surge.sh

dlabrecq commented:
Looks like we have two PRs for the same issue?
https://github.com/patternfly/patternfly-react/pull/3232

dlabrecq commented on packages/patternfly-4/react-core/src/demos/PaginationTable.md:
```diff
@@ -50,17 +70,50 @@ class ComplexPaginationTableDemo extends React.Component {
+            )
+          },
+        ]
+      }]
```
How do users see the empty state? Do we need to edit the code to remove rows?

boaz0 replied:
we should probably add a filter to the table and if a filter is set show empty state.

dlabrecq commented on packages/patternfly-4/react-core/src/demos/PaginationTable.md:
```diff
@@ -50,17 +70,50 @@ class ComplexPaginationTableDemo extends React.Component {
           </Table>
         )}
-        {loading && <center><Title size="3xl">Please wait while loading data</Title></center>}
+        {loading && <center><Spinner size="xl"/></center>}
```
I'm not seeing a spinner. I briefly see the empty state prior to loading the table rows. Is there a way to better see this?

mcarrano commented:
@nicolethoen for some reason I don't seem to be able to access the preview link. It says "project not found."

nicolethoen commented:
@dlabrecq @boaz0 I can work to add a way to force view these states.

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-3294.surge.sh

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-3294.surge.sh

nicolethoen commented:
@mcarrano those changes should be no problem! There are no filters on the table though, so the 'Clear All Filters' action would be a dead action, so do we still want it? I removed it intentionally for just that reason.

mcarrano commented:
That's an interesting question @nicolethoen . I think it's important to have the Clear filters action since this message would only really appear as the result of a filter. But you are right that this would not really do anything in the current demo. Would it take much to add a simple filter to this, like just a Select list with some values?

nicolethoen commented:
@mcarrano we have a filterable table demo... would it need to also use the experimental data toolbar to be implemented correctly? seems out of scope for this demo. The only legitimate reason there would be no data to display is if no data is returned from the API. It'd be more of an error.

mcarrano commented:
Yes, that's true @nicolethoen . But this is a slightly different empty state pattern in that case. I will try to set up some time to sync with you on what to put here.

nicolethoen commented:
Note: radio buttons were removed - it was decided they were muddying context for the user a little bit. And 'Empty State' was replaced with more of an 'Error state' that is not demonstrated intentionally as part of the demo - but if there is an ever some unknown error retrieving data, it will fail gracefully. Any anyone curious about error handling in this case can look at how it's handled in the code. It was decided this demo is not going to highlight the error state, because it is not an Error State demo, it is a Pagination Table demo.

nicolethoen commented:
@dlabrecq I dont think that the Pagination Demo is the place to demonstrate the empty state or error state of a table. It is simply for demonstrating pagination. The other demos don't explicitly demonstrate empty states or error states either. 

Before the change, as data was being loaded and paginated, you saw a message that quickly flashed `Please wait while loading data` before it disappeared. Now you see the spinner instead. 

And we've added an error state if no data is retrieved, which should never happen unless there is a connectivity issue - so I added the error state so that any issues that arise will fail gracefully using a Patternfly pattern. So it's not explicitly demonstrable, but is more there as it should if anyone wants to implement Pagination in their table correctly. I think that is following the demo precedent we've established.

These changes were made after talking to @mcarrano and Content strategist Margot

nicolethoen commented:
I'll also add that the Filterable Table demo does demonstrate an empty state when all options have been filtered out. That seems to be a more relevant place to demonstrate this.

dlabrecq commented:
I'm just going by what issue #3184 is asking for. If the functionality is not appropriate for the pagination demo, then perhaps we should not modify this particular demo? Or, leave the issue open?

Regardless, my issue is that I don't even see the spinner.

nicolethoen commented:
Your machine must be processing the data too fast for you to see it. @mcarrano would it be appropriate to add a 'View Loading State" checkbox to force the table to stay in a loading state?

dlabrecq commented:
How about a timeout that loads the table after a second or two? Then, there is no need for a checkbox

mcarrano commented:
@nicolethoen @dlabrecq I'm good with either of those approaches. It was the inclusion of the empty state here that was causing me the most confusion. I'm wondering if we should have a demo that shows the various types of empty states in context as there are lots of different reasons to get an empty state. Do you think that would be useful?

nicolethoen commented:
@dlabrecq I updated the attached issue to reflect the changes made in this PR. And I opened a new  issue to document the missing demos. #3330

nicolethoen commented:
Any other thoughts @dlabrecq or @mcarrano? Can we get this PR in soon :)

mcarrano commented:
@nicolethoen It makes sense to me to create these in a separate but related demo. I've left a comment to see if we can get that om the roadmap.

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@1.1.33
 - @patternfly/react-core@3.122.3
 - @patternfly/react-docs@4.16.39
 - @patternfly/react-inline-edit-extension@2.13.4
 - demo-app-ts@3.11.3
 - @patternfly/react-table@2.24.36
 - @patternfly/react-topology@2.11.22
 - @patternfly/react-virtualized-extension@1.3.35

Thanks for your contribution! :tada:
